packages/ckeditor5-engine/src/view/matcher.js

```diff
@@ -242,71 +244,183 @@ function matchName( pattern, name ) {
 // @returns {Array|null} Returns array with matched attribute names or `null` if no attributes were matched.
-function matchAttributes( patterns, element ) {
+function matchPatterns( patterns, attributes ) {
+	const attributeKeys = attributes.map( ( [ key ] ) => key );
 	const match = [];
 
-	for ( const name in patterns ) {
-		const pattern = patterns[ name ];
-
-		if ( element.hasAttribute( name ) ) {
-			const attribute = element.getAttribute( name );
-
-			if ( pattern === true ) {
-				match.push( name );
-			} else if ( pattern instanceof RegExp ) {
-				if ( pattern.test( attribute ) ) {
-					match.push( name );
-				} else {
-					return null;
-				}
-			} else if ( attribute === pattern ) {
-				match.push( name );
-			} else {
-				return null;
+	if ( patterns === true ) {
+		if ( attributeKeys.length ) {
+			return attributeKeys;
+		} else {
+			return null;
+		}
+	} else if ( patterns instanceof RegExp ) {
+		attributeKeys.forEach( attribute => {
+			if ( patterns.test( attribute ) ) {
+				match.push( attribute );
 			}
+		} );
+
+		if ( match.length ) {
+			return match;
 		} else {
 			return null;
 		}
 	}
 
+	patterns = normalizePatterns( patterns ) || [];
+
+	patterns.forEach( ( { key: patternKey, value: patternValue } ) => {
+		attributes.forEach( ( [ attributeKey, attributeValue ] ) => {
+			if (
+				isAttributeKeyMatched( patternKey, attributeKey ) &&
+				isAttributeValueMatched( patternValue, attributeValue )
+			) {
+				match.push( attributeKey );
+			}
+		} );
+	} );
+
+	// Return null when there was no matches or we didn't match all patterns.
+	if ( !match.length || match.length != patterns.length ) {
+		return null;
+	}
+
 	return match;
 }
 
+// Bring all the possible pattern forms to an array of objects with `key` and `value` property.
+// For example:
+//
+//	{
+//		src: /https:.*/
+//	}
+//
+// to:
+//
+//	{
+//		key: 'src',
+//		value: /https:.*/
+//	}
+// @param {Object|Array} patterns
+// @returns {Array|null} Returns an array of objects or null if provided patterns were not in an expected form.
+function normalizePatterns( patterns ) {
+	if ( Array.isArray( patterns ) ) {
+		return patterns.map(
+			pattern => {
+				// eslint-disable-next-line dot-notation
+				if ( isPlainObject( pattern ) && pattern[ 'key' ] && pattern[ 'value' ] ) {
+					return pattern;
+				}
+
+				// Assume the pattern is either String or RegExp.
+				return { key: pattern, value: true };
+			}
+		);
+	} else if ( isPlainObject( patterns ) ) {
+		return Object.entries( patterns ).map(
+			( [ patternKey, patternValue ] ) => ( { key: patternKey, value: patternValue } )
+		);
+	}
+
+	return null;
+}
+
+// @param {String|RegExp} patternKey A pattern representing attribute key we want to match.
+// @param {String} attributeKey An actual attribute key (e.g. `'src'`, `'background-color'`, `'ck-widget'`) we're testing against pattern.
+// @returns {Boolean}
+function isAttributeKeyMatched( patternKey, attributeKey ) {
```

Comment:
Maybe just `matchAttributeKey`?

Reply:
After small talk let's go with the current naming as it better indicates that we are expecting boolean value.